Support event relations on all post message functions

nvrWhere opened the pull request:
Title

nvrWhere committed:
Support event relations on all post message functions

KitsuneRal commented on Quotient/room.h:
```diff
@@ -746,15 +746,16 @@ public Q_SLOTS:
-    QString postMessage(const QString& plainText, MessageEventType type);
-    QString postPlainText(const QString& plainText);
+    QString postMessage(const QString& plainText, MessageEventType type, std::optional<EventRelation> relatesTo = std::nullopt);
+    QString postPlainText(const QString& plainText, std::optional<EventRelation> relatesTo = std::nullopt);
```
Now that I look at this code I just realised there may be a small problem: these methods are meant to be invocable from QML (they are incorrectly marked as "slots" they should be `Q_INVOKABLE` instead) and `std::optional` is not mapped to QML so QML will never be able to use that new last argument. I don't think we call those methods from QML in Quaternion and NeoChat but I wouldn't want to rule out the possibility. Any ideas how to work around it are welcome; or we could just accept this limitation and go with it.

nvrWhere replied:
So the reason I never considered it was that I'm not sure they should be called from QML. As you've correctly noted we don't in NeoChat for 2 main reasons:
1. In order to implement features like mentions and other advanced things we need to process the string
2. All well behaving clients should check the string for improper html (see https://spec.matrix.org/v1.8/client-server-api/#mroommessage-msgtypes) which also needs some preprocessing

All this means as far as I see it is that unless someone plans to fork QTextDocument and extend it with all this (other KDE apps have so it's not impossible), these functions shouldn't be called from QML.

That being said if we want to leave this on the table we could do it with a QString replyEventId and make the EventRelation ourselves. I'll then just stick some heavy warnings in the documentation

TobiasFella replied:
We could also change the functions to have overloads that take directly take a relation and one that doesn't take a relation. But if we don't even need them from qml for now, i would also be fine with just not being able to use them from qml

KitsuneRal replied:
Sorry, this has slipped through the cracks. Thanks for the reasons, they look solid at the current state of the library. Eventually, mentions will have to be processed by the library; and I have an idea to upstream the current HTML preprocessing code from Quaternion, too. But all that is unlikely to happen in immediate future so it's probably best to just drop invocability from this methods.

nvrWhere committed:
Make all the functions non-callable from QML and remove postMessage because we probably shouldn't allow someone to send a message with an arbritary type but not the correct content. This is replaced by postEmote and postNotice the two types that are probably valid from the original postMessage function.

nvrWhere commented:
So I updated the functions to make them not invokable as discussed.  Also since we're breaking API i messed with them a little. 

I removed `postMessage` because we probably shouldn't allow someone to send a message with an arbitrary type but not the correct content. This is replaced by `postEmote` and `postNotice`  which along with `postHtmlText` are the types that are probably valid from the original `postMessage` function.

KitsuneRal commented on Quotient/room.h:
```diff
@@ -721,8 +721,28 @@ class QUOTIENT_API Room : public QObject {
         return post(makeEvent<EvT>(std::forward<ArgTs>(args)...));
     }
 
+    /// Send a plain text message
+    QString postPlainText(const QString& plainText, std::optional<EventRelation> relatesTo = std::nullopt);
+
+    /// Send a rich text message
+    QString postHtmlText(const QString& plainText, const QString& html, std::optional<EventRelation> relatesTo = std::nullopt);
+
+    /// Send a m.emote message
+    QString postEmote(const QString& plainText, std::optional<const QString> html = std::nullopt, std::optional<EventRelation> relatesTo = std::nullopt);
+
+    /// Send an m.notice message
+    QString postNotice(const QString& plainText, std::optional<const QString> html = std::nullopt, std::optional<EventRelation> relatesTo = std::nullopt);
+
+    /// Send a file with the given content
```
```suggestion
    //! Send a plain text message
    QString postPlainText(const QString& plainText, std::optional<EventRelation> relatesTo = std::nullopt);

    //! Send a rich text message
    QString postHtmlText(const QString& plainText, const QString& html, std::optional<EventRelation> relatesTo = std::nullopt);

    //! Send a m.emote message
    QString postEmote(const QString& plainText, std::optional<const QString> html = std::nullopt, std::optional<EventRelation> relatesTo = std::nullopt);

    //! Send an m.notice message
    QString postNotice(const QString& plainText, std::optional<const QString> html = std::nullopt, std::optional<EventRelation> relatesTo = std::nullopt);

    //! Send a file with the given content
```

KitsuneRal commented on Quotient/room.h:
```diff
@@ -721,8 +721,28 @@ class QUOTIENT_API Room : public QObject {
+                     std::unique_ptr<EventContent::FileContentBase> fileContent,
+                     std::optional<EventRelation> relatesTo = std::nullopt);
+
+    /// Send the given Json as a message
+    QString postJson(const QString& matrixType, const QJsonObject& eventContent);
+
+    /// Send a reaction on a given event with a given key
+    QString postReaction(const QString& eventId, const QString& key);
```
```suggestion
    //! Send the given JSON as a message
    QString postJson(const QString& matrixType, const QJsonObject& eventContent);

    //! Send a reaction on a given event with a given key
    QString postReaction(const QString& eventId, const QString& key);
```

KitsuneRal commented on Quotient/room.h:
```diff
@@ -721,8 +721,28 @@ class QUOTIENT_API Room : public QObject {
+    /// Send a plain text message
+    QString postPlainText(const QString& plainText, std::optional<EventRelation> relatesTo = std::nullopt);
+
+    /// Send a rich text message
+    QString postHtmlText(const QString& plainText, const QString& html, std::optional<EventRelation> relatesTo = std::nullopt);
+
+    /// Send a m.emote message
+    QString postEmote(const QString& plainText, std::optional<const QString> html = std::nullopt, std::optional<EventRelation> relatesTo = std::nullopt);
+
+    /// Send an m.notice message
+    QString postNotice(const QString& plainText, std::optional<const QString> html = std::nullopt, std::optional<EventRelation> relatesTo = std::nullopt);
```
Might I suggest making it a single templated inline member function, with the non-type template parameter defaulting to `MessageEventType::Text`? You can static-assert the correct msgtype in it. And now that I think of it, perhaps `postPlainText()` and `postHtmlText()` can be merged into a single `postText()` function, because the original distinction between the two was made solely for the purpose of accessibility from QML, back in Spectral times.

KitsuneRal commented on Quotient/room.cpp:
```diff
@@ -52,6 +52,7 @@
 #include "events/roomcanonicalaliasevent.h"
 #include "events/roomcreateevent.h"
 #include "events/roommemberevent.h"
+#include "events/roommessageevent.h"
```
```suggestion
```
It's already included from `room.h` (but yeah, I slowly start thinking that include-what-you-use might ultimately be a good practice... but I'm not there yet).

KitsuneRal reviewed (commented):
The template suggestion is just a suggestion, open for discussion.

nvrWhere committed:
Comment updates

KitsuneRal commented on Quotient/room.cpp:
```diff
@@ -293,6 +293,25 @@ class Q_DECL_HIDDEN Room::Private {
+    // template<MessageEventType msgType>
+    template<MessageEventType type>
+    QString postAllText(const QString& plainText,
+                           std::optional<const QString> html,
```
```suggestion
                           const std::optional<QString>& html,
```

KitsuneRal commented on Quotient/room.cpp:
```diff
@@ -293,6 +293,25 @@ class Q_DECL_HIDDEN Room::Private {
     const PendingEventItem& sendEvent(RoomEventPtr&& event);
 
+    // template<MessageEventType msgType>
+    template<MessageEventType type>
```
```suggestion
    template<MessageEventType type = MessageEventType::Text>
```

KitsuneRal commented on Quotient/room.cpp:
```diff
@@ -293,6 +293,25 @@ class Q_DECL_HIDDEN Room::Private {
 
+    // template<MessageEventType msgType>
+    template<MessageEventType type>
+    QString postAllText(const QString& plainText,
```
Any reason not to make it a part of `Room` interface? With the default template parameter value suggested above, you could just have a single `postText()` method template instead of having `postText()` and `postAllText()` separately; and implement `postNotice()` and `postEmote()` in terms of `postText()`.

nvrWhere replied:
The answer is I didn't have any particular reason other than not thinking of it. It's now just a single method.

nvrWhere committed:
Further comments

KitsuneRal commented:
Ftr (my own, in particular): this PR targets 0.10, not 0.9

nvrWhere commented:
> Ftr (my own, in particular): this PR targets 0.10, not 0.9

Indeed that was my assumption since there are API breaks

KitsuneRal committed:
Merge branch 'dev' into nvrwhere/relatesTo_post

usta commented:
@KitsuneRal https://github.com/quotient-im/libQuotient/blob/700d54733ba3b0c081e4cb552de02b274826be7e/quotest/quotest.cpp#L908

Looks like you forget one postPlainText ^^

KitsuneRal commented:
It's already being fixed

usta commented:
now it is fixed thanks :)
